File: src/simulation/space/resource/MutableResourcePack.java
package simulation.space.resource;

import kotlin.Pair;
import simulation.space.Tile;
import simulation.space.resource.tag.ResourceTag;

import java.util.*;

public class MutableResourcePack extends ResourcePack{ //TODO subclass which stores all instances of the same Resource on different Tiles

    public MutableResourcePack(Collection<Resource> resources) {
        super(resources);
    }

    public MutableResourcePack() {
        this(Collections.emptyList());
    }

    public ResourcePack getResourceAndRemove(Resource resource) {
        ResourcePack _r = getResource(resource);
        _r.getResources().forEach(this::remove);
        return _r;
    }

    public MutableResourcePack getResourcesWithTagPartIsBigger(ResourceTag tag, int ceiling) {
        return getPart(getAllResourcesWithTag(tag), ceiling);
    }

    public MutableResourcePack getResourcePart(Resource resource, int ceiling) {
        return getPart(getResource(resource), ceiling);
    }

    private MutableResourcePack getPart(ResourcePack pack, int amount) {
        MutableResourcePack result = new MutableResourcePack();
        int counter = 0;
        for (Resource resource : pack.getResources()) {
            if (counter >= amount) {
                break;
            }
            result.add(resource);
            counter += resource.getAmount();
        }
        return result;
    }

    public Pair<Integer, List<Resource>> getAmountOfResourcesWithTagAndErase(ResourceTag tag, int amount) {//TODO looks bad
        Collection<Resource> _r = getAllResourcesWithTag(tag).getResources();
        List<Resource> result = new ArrayList<>();
        int counter = 0;
        for (Resource resource : _r) {
            if (counter >= amount) {
                break;
            }
            result.add(resource);
            counter += resource.getAmount();
        }
        result.forEach(this::remove);
        return new Pair<>(counter, result);
    }

    public void disbandOnTile(Tile tile) {
        getResources().stream()
                .filter(resource -> resource.getGenome().isMovable())
                .forEach(tile::addDelayedResource);
        resourcesMap.clear();
    }

    public void add(Resource resource) {
        internalAdd(resource);
    }

    public void addAll(ResourcePack resourcePack) {
        resourcePack.getResources().forEach(this::add);
    }

    public void addAll(Collection<Resource> resources) {
        resources.forEach(this::add);
    }

    public void remove(Resource resource) {
        resourcesMap.remove(resource);
    }

    public void removeAll(Collection<Resource> resources) {
        resources.forEach(this::remove);
    }

    public void removeAll(ResourcePack pack) {
        pack.getResources().forEach(this::remove);
    }

    public void destroyAllResourcesWithTag(ResourceTag tag) {
        ResourcePack result = getAllResourcesWithTag(tag);
        removeAll(result);
        result.getResources().forEach(resource -> resource.setAmount(0));
    }

    @Override
    public String toString() {
        StringBuilder stringBuilder = new StringBuilder();
        for (Resource resource : getResources()) {
            stringBuilder.append(resource.getFullName()).append(" ")
                    .append(resource.getAmount())
                    .append("; ")
                    .append("\n");
        }
        return stringBuilder.toString();
    }
}

File: src/simulation/space/resource/MutableResourcePack.kt
package simulation.space.resource

import simulation.space.tile.Tile
import simulation.space.resource.tag.ResourceTag
import java.util.*
import kotlin.math.min

class MutableResourcePack(resources: Collection<Resource> = emptyList()) : ResourcePack(resources) {
    constructor(pack: ResourcePack) : this(pack.resources)

    fun getResourceAndRemove(resource: Resource): ResourcePack {
        val innerResource = getResource(resource)
        innerResource.resources.forEach { this.remove(it) }
        return innerResource
    }

    fun getResourcePartAndRemove(resource: Resource, amount: Int): ResourcePack {
        val innerResource = getResource(resource)
        val result = MutableResourcePack()
        var resultAmount = 0
        for (res in innerResource.resources) {
            val resAmount = min(res.amount, amount - resultAmount)
            resultAmount += resAmount
            if (resAmount == res.amount) {
                result.add(res)
                remove(res)
            } else {
                result.add(resource.getCleanPart(resAmount))
            }
        }
        return result
    }

    fun getResourcePart(resource: Resource, ceiling: Int): MutableResourcePack =
            getPart(getResource(resource), ceiling)

    private fun getPart(pack: ResourcePack, amount: Int): MutableResourcePack {
        val result = MutableResourcePack()
        var counter = 0
        for (resource in pack.resources) {
            if (counter >= amount)
                break
            result.add(resource)
            counter += resource.amount
        }
        return result
    }

    fun getAmountOfResourcesWithTagAndErase(tag: ResourceTag, amount: Int): Pair<Int, List<Resource>> { //TODO looks bad
        val innerResource: Collection<Resource> = getResources(tag).resources
        val result: MutableList<Resource> = ArrayList()
        var counter = 0
        for (resource in innerResource) {
            if (counter >= amount) {
                break
            }
            result.add(resource)
            counter += resource.amount
        }
        result.forEach { this.remove(it) }
        return Pair<Int, List<Resource>>(counter, result)
    }

    fun disbandOnTile(tile: Tile) {
        tile.addDelayedResources(resources.filter {  it.genome.isMovable })
        resourceMap.clear()
    }

    /**
     * Returns true only if the Resource was merged
     */
    fun add(resource: Resource): Boolean = internalAdd(resource)

    fun addAll(resources: Collection<Resource>) = resources.forEach { add(it) }

    fun addAll(resourcePack: ResourcePack) = addAll(resourcePack.resources)

    fun remove(resource: Resource) = resourceMap.remove(resource)

    fun removeAll(resources: Collection<Resource>) = resources.forEach { this.remove(it) }

    fun removeAll(pack: ResourcePack) = removeAll(pack.resources)

    fun destroyAllResourcesWithTag(tag: ResourceTag) {
        val result = getResources(tag)
        removeAll(result)
        result.resources.forEach { it.destroy() }
    }
}